docs: Update security.md to include Trusted Types (#40142)
angular/angular: aio/content/guide/security.md

# Security

This topic describes Angular's built-in
protections against common web-application vulnerabilities and attacks such as cross-site
scripting attacks. It doesn't cover application-level security, such as authentication and authorization.

For more information about the attacks and mitigations described below, see [OWASP Guide Project](https://www.owasp.org/index.php/Category:OWASP_Guide_Project).

You can run the <live-example></live-example> in Stackblitz and download the code from there.

<div class="callout is-important">

> <header>Reporting vulnerabilities</header>

Reviewer: ```suggestion
{@a report-issues}
<header>Reporting vulnerabilities</header>
```

To report vulnerabilities in Angular itself, email us at [security@angular.io](mailto:security@angular.io).

For more information about how Google handles security issues, see [Google's security
philosophy](https://www.google.com/about/appsecurity/).

</div>

<div class="callout is-helpful">

> <header>Best practices</header>

Reviewer: ```suggestion
{@a best-practices}
<header>Best practices</header>
```

DOM allow code execution, for example, `<img onerror="...">` and `<a href="javascript:...">`. If
attacker-controlled data enters the DOM, expect security vulnerabilities.

{@a trusted-types}
### Implementing Trusted Types

One recommended way to help secure your applications from cross-site scripting attacks is to implement Trusted Types. Trusted Types is a [web platform](https://en.wikipedia.org/wiki/Web_platform)
feature that can help you prevent cross-site scripting attacks by enforcing
safer coding practices. Trusted Types can also help simplify the auditing of application code.

> Support for Trusted Types is an opt-in feature. To implement Trusted Types for your application, you must configure your application's HTTP headers with one of the following Angular policies:

Reviewer 2: I think this can use a little bit of rephrasing. The thing is that *support* for Trusted Types is not really something that is opt-in. If your browser supports Trusted Types, your application code can make use of the Trusted Types API (which we just refer to as Trusted Types) right away, and you don't even need to configure any headers to do so. And indeed, as of v11.1, Angular will use Trusted Types under the hood right away in compatible browsers, irrespective of which headers are set. The feature that *is* opt-in is whether Trusted Types are *enforced* in your application -- this is what the headers configure. When Trusted Types are enforced, the browser blocks any use of sensitive DOM APIs unless the input to that API is derived from an appropriate Trusted Types policy.

With that in mind, consider changing:
- "Support for Trusted Types is an opt-in feature" to something like "Angular is compatible with Trusted Types", since before v11.1 enforcing Trusted Types would actually break Angular applications.
- "To implement Trusted Types for your application" to something like "To enforce Trusted Types in your application".

Reviewer: I agree with Bjakri on this one. I can draft the restructured version of this content if that makes it easier.

> * `angular`. This policy is the default Trusted Type policy used by the Angular [Sanitizer](api/core/Sanitizer). Any inline template values or content sanitized by Angular is treated as safe by this policy.

Reviewer 2: Consider changing "This policy is the default Trusted Type policy used by the Angular [Sanitizer](api/core/Sanitizer)" to something like "This policy is used in security-reviewed code that is internal to Angular, and is required for Angular to function when Trusted Types are enforced.".

> * `angular#unsafe-bypass`. This policy is for applications that use Angular's [DomSanitizer](api/platform-browser/DomSanitizer). Content sanitized with the DomSanitizer is treated as safe.

Reviewer 2: I would classify DomSanitizer as also being handled by the `angular` policy. Instead, the `angular#unsafe-bypass` policy is for applications that use any of the `bypassSecurityTrust...` functions or a custom DomSanitizer, where application authors are responsible for the security of their code. While most applications currently require this policy, it will be discouraged in favour of a more direct use of Trusted Types in a future version of Angular.

* `angular#unsafe-jit`. This policy is used by the [JIT compiler](api/core/Compiler). You must enable this policy if your application interacts directly with the JIT compiler or is running in JIT mode using the [platform browser dynamic](api/platform-browser-dynamic/platformBrowserDynamic).

You should configure the HTTP headers for Trusted Types in the following locations:

* Production serving infrastructure, for production environments

> * Angular CLI (`ng serve`), using the `angular.json` file, for local development and end-to-end testing

Reviewer 2: ... using the `headers` property in the `angular.json` file ...

I think it makes sense to call this out specifically since the `headers` property/setting is new.

> * Karma (`ng test`), using `karma.config.js` for unit testing

Reviewer 2: ... using the `customHeaders` property in `karma.config.js` ...

The following is an example of a header specifically configured for Trusted Types and Angular:

```

> Content-Security-Policy: trusted-types 'angular' 'angular#unsafe-bypass';
>   required-trusted-types-for 'script';

Reviewer: Let's remove this new line from here - it results in the value being rendered in a way that suggest that the new line is required: 

```suggestion
Content-Security-Policy: trusted-types 'angular' 'angular#unsafe-bypass'; required-trusted-types-for 'script';
```

```

The following an example of header specifically configured for Trusted Types and Angular applications using JIT:

```

> Content-Security-Policy: trusted-types 'angular' 'angular#unsafe-bypass' 'angular#unsafe-jit';
>   required-trusted-types-for 'script';

Reviewer: ```suggestion
Content-Security-Policy: trusted-types 'angular' 'angular#unsafe-bypass' 'angular#unsafe-jit'; required-trusted-types-for 'script';
```

> ```

Reviewer 2: Since enabling Trusted Types in enforcing mode can actually break applications in production, I think it would be a good idea to add a warning callout with something like "Make sure you test your application with report-only Trusted Types (Content-Security-Policy-Report-Only) before enforcing Trusted Types in production".

Reviewer 3: Where should this warning go? Could you clarify?

[Web Fundamentals guide](https://developers.google.com/web/fundamentals/security/csp) on the
Google Developers website.

{@a trusted-types}

> ### Implementing Trusted Types

Reviewer: "Implementing" here is problematic as it gives impression that the developers will have to implement the trusted types. In reality, they reconfigure their application in a way that tells the browser to enforce trusted types. Consider changing the title to just "Trusted Types" or one of:
- Leveraging Trusted Types
- Trusted Types support
- Using Trusted Types

> One recommended way to help secure your applications from cross-site scripting attacks is to use Trusted Types. Trusted Types is a [web platform](https://en.wikipedia.org/wiki/Web_platform)

Reviewer: ```suggestion
The most recent and recommended way to help secure your applications from cross-site scripting attacks is to use Trusted Types. Trusted Types is a [web platform](https://en.wikipedia.org/wiki/Web_platform)
```

feature that can help you prevent cross-site scripting attacks by using security policies to restrict access to sensitive DOM APIs. Trusted Types can also help simplify the auditing of application code.

To enforce Trusted Types for your application, you must configure your application's HTTP headers with one of the following Angular policies:

> * `angular`. This policy is used in security-reviewed code that is internal to Angular, and is required for Angular to function when Trusted Types are enforced. Any inline template values or content sanitized by Angular is treated as safe by this policy.

Reviewer: ```suggestion
* `angular`. This policy is used in security-reviewed code that is internal to Angular, and is required for Angular to function when Trusted Types are enforced. Any inline template values or content sanitized by Angular is marked as safe and trusted using this policy.
```

Reviewer 3: Rewritten. PTAL?

> * `angular#unsafe-bypass`. This policy is used for applications that use any of the methods in Angular's [DomSanitizer](api/platform-browser/DomSanitizer) that bypasses security, such as [`bypassSecurityTrustHtml`](/api/platform-browser/DomSanitizer#bypasssecuritytrusthtml). Content sanitized with the DomSanitizer is treated as safe.

Reviewer: ```suggestion
* `angular#unsafe-bypass`. This policy is used by methods in Angular's [DomSanitizer](api/platform-browser/DomSanitizer) that bypasses security, such as [`bypassSecurityTrustHtml`](/api/platform-browser/DomSanitizer#bypasssecuritytrusthtml). Any application that uses these `bypass*` methods will need to enable this policy. This policy is provided primarily for existing code that already uses the bypass methods. Once Trusted Types gain sufficient adoption by browser vendors, these bypass apis will be deprecated and eventually removed. New code should instead [create a custom policy](https://web.dev/trusted-types/#create-a-trusted-type-policy) specific to the application and library, and create data used in sensitive contexts using this custom policy.
```

Reviewer 3: We do not document the future in our content, so we can't make statements such as "After x occurs, we will do y."

* `angular#unsafe-jit`. This policy is used by the [JIT compiler](api/core/Compiler). You must enable this policy if your application interacts directly with the JIT compiler or is running in JIT mode using the [platform browser dynamic](api/platform-browser-dynamic/platformBrowserDynamic).

You should configure the HTTP headers for Trusted Types in the following locations:

* Production serving infrastructure, for production environments
* Angular CLI (`ng serve`), using the `headers` property in the `angular.json` file, for local development and end-to-end testing
* Karma (`ng test`), using the `customHeaders` property in `karma.config.js` for unit testing

The following is an example of a header specifically configured for Trusted Types and Angular:

```
"Content-Security-Policy: trusted-types angular angular#unsafe-bypass;
  require-trusted-types-for 'script';"

> ```

Reviewer: Can we split this into two parts - one for the default config, and one for the cases when bypass apis are used? Something like this:

----

The following is an example of a header specifically configured for Trusted Types and Angular:

```
Content-Security-Policy: trusted-types angular; require-trusted-types-for 'script';
```

The following is an example of a header specifically configured for Trusted Types and Angular applications that use `DomSanitizer#bypass*` APIs:

```
Content-Security-Policy: trusted-types angular angular#unsafe-bypass; require-trusted-types-for 'script';
```

The following is an example of a header specifically configured for Trusted Types and Angular applications using JIT:

```

> "Content-Security-Policy: trusted-types angular angular#unsafe-bypass angular#unsafe-jit;
>   require-trusted-types-for 'script';"

Reviewer: There cannot be a new line within the header value and the quotes should be removed as well since all of this will be rendered as code block:
```suggestion
Content-Security-Policy: trusted-types angular angular#unsafe-bypass angular#unsafe-jit; require-trusted-types-for 'script';
```

<div class="callout is-helpful">

Trusted Types might not yet be available in all browsers your application targets. In the case your Trusted-types-enabled application runs in a browser that doesn't support trusted types, the functionality of the application will be preserved, and your application will be guarded against XSS via Angular's DomSanitizer. See [caniuse.com/trusted-types](https://caniuse.com/trusted-types) for the current browser support.

</div>

To enforce Trusted Types for your application, you must configure your application's web server to emit HTTP headers with one of the following Angular policies:

* `angular` - This policy is used in security-reviewed code that is internal to Angular, and is required for Angular to function when Trusted Types are enforced. Any inline template values or content sanitized by Angular is treated as safe by this policy.
* `angular#unsafe-bypass` - This policy is used for applications that use any of the methods in Angular's [DomSanitizer](api/platform-browser/DomSanitizer) that bypass security, such as `bypassSecurityTrustHtml`. Any application that uses these methods must enable this policy.

> * `angular#unsafe-jit`. This policy is used by the [JIT compiler](api/core/Compiler). You must enable this policy if your application interacts directly with the JIT compiler or is running in JIT mode using the [platform browser dynamic](api/platform-browser-dynamic/platformBrowserDynamic).

Reviewer: +1

You should configure the HTTP headers for Trusted Types in the following locations:

* Production serving infrastructure

> * Angular CLI (`ng serve`), using the `headers` proprty in the `angular.json` file, for local development and end-to-end testing

Reviewer: ```suggestion
* Angular CLI (`ng serve`), using the `headers` property in the `angular.json` file, for local development and end-to-end testing
```

* Karma (`ng test`), using the `customHeaders` property in the `karma.config.js` file, for unit testing

The following is an example of a header specifically configured for Trusted Types and Angular:

```

> Content-Security-Policy: trusted-types 'angular'; required-trusted-types-for 'script';

Reviewer: yup, I think this is a copy&paste mistake. @aikidave can you please update this part? thanks!

```

The following is an example of a header specifically configured for Trusted Types and Angular applications that use any of the methods in Angular's [DomSanitizer](api/platform-browser/DomSanitizer) that bypasses security.

```
Content-Security-Policy: trusted-types angular angular#unsafe-bypass; require-trusted-types-for 'script';

> ```

Reviewer: I don't recall if we support turning of syntax highlighting for triple backticks. @aikidave, perhaps consider using `<code>` block?

An alternative to the AOT compiler is the JIT compiler which compiles templates to executable template code within the browser at runtime. Angular trusts template code, so dynamically generating templates and compiling them, in particular templates containing user data, circumvents Angular's built-in protections and is a security anti-pattern. For information about dynamically constructing forms in a safe way, see the [Dynamic Forms](guide/dynamic-form) guide.

### Server-side XSS protection

HTML constructed on the server is vulnerable to injection attacks. Injecting template code into an Angular application is the same as injecting executable code into the application: it gives the attacker full control over the application. To prevent this, use a templating language that automatically escapes values to prevent XSS vulnerabilities on the server. Don't generate Angular templates on the server side using a templating language; doing this carries a high risk of introducing template-injection vulnerabilities.

> ## HTTP-level vulnerabilities

Reviewer: ```suggestion
{@a http}
## HTTP-level vulnerabilities
```

Angular has built-in support to help prevent two common HTTP vulnerabilities, cross-site request
forgery (CSRF or XSRF) and cross-site script inclusion (XSSI). Both of these must be mitigated primarily
on the server side, but Angular provides helpers to make integration on the client side easier.

> ### Cross-site request forgery

Reviewer: ```suggestion
{@a xsrf}
### Cross-site request forgery
```

See also Dave Smith's easy-to-understand
[talk on XSRF at AngularConnect 2016](https://www.youtube.com/watch?v=9inczw6qtpY "Cross Site Request Funkery Securing Your Angular Apps From Evil Doers").

> ### Cross-site script inclusion (XSSI)

Reviewer: ```suggestion
{@a xssi}
### Cross-site script inclusion (XSSI)
```

For more information, see the XSSI section of this [Google web security blog
post](https://security.googleblog.com/2011/05/website-security-for-webmasters.html).

> ## Auditing Angular applications

Reviewer: ```suggestion
{@a code-review}
## Auditing Angular applications
```

For more information, see the XSSI section of this [Google web security blog
post](https://security.googleblog.com/2011/05/website-security-for-webmasters.html).

> {@ code-review}

Reviewer: ```suggestion
{@a code-review}
```
